modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/SurveyEventsFormListener.java

```diff
@@ -0,0 +1,262 @@
+                ) {
+                    surveyEventsForm = referencedNode;
+                }
+            }
+        }
+
+        if (surveyEventsForm != null) {
+            // Found a survey events form: Link to it
+            this.linkUtils.addLink(form, surveyEventsForm, "belongsToSurvey", "Belongs to Survey");
+        }
+    }
+
+    private List<String> getVisitClinicQuestionnaires(Session session, Node subject)
+        throws RepositoryException
+    {
+        // Get relevant nodes on the visit information form
+        Node visitInformationQuestionnaire = session.getNode(VISIT_INFORMATION_PATH);
+        Node visitClinicQuestion = visitInformationQuestionnaire.getNode("clinic");
+        Node visitInformationForm = getVisitInformationForm(subject, visitInformationQuestionnaire);
+
+        if (visitInformationForm != null) {
+            // Get the list of questionnaires associated with the current clinic
+            Node clinic = this.formUtils.getAnswer(visitInformationForm, visitClinicQuestion);
+            if (clinic.hasProperty("value")) {
+                return getClinicQuestionnaires(session, clinic.getProperty("value").getString());
+            }
+        }
+        return new ArrayList<String>();
+    }
+
+    private List<String> getClinicQuestionnaires(Session session, String clinic)
+        throws RepositoryException
+    {
+        Node clinicMapping = session.getNode(clinic);
+        Node survey = session.getNode("/Survey/" + clinicMapping.getProperty("survey").getString());
+        List<String> surveyFormIds = new LinkedList<>();
+        NodeIterator surveyForms = survey.getNodes();
+        while (surveyForms.hasNext()) {
+            Node surveyForm = surveyForms.nextNode();
+            if ("cards:QuestionnaireRef".equals(surveyForm.getPrimaryNodeType().getName())) {
+                surveyFormIds.add(surveyForm.getProperty("questionnaire").getNode().getIdentifier());
+            }
+        }
+        return surveyFormIds;
+    }
+
+    private boolean isVisitInformationForm(final Node form)
+    {
+        try {
+            return VISIT_INFORMATION_PATH.equals(this.formUtils.getQuestionnaire(form).getPath());
+        } catch (RepositoryException e) {
+            LOGGER.warn("Failed check if form is Visit information form: {}", e.getMessage(), e);
+            return false;
+        }
+    }
+
+    private boolean isSurveyEventsQuestionnaire(final Node questionnaire)
+    {
+        try {
+            return questionnaire != null && SURVEY_EVENTS_PATH.equals(questionnaire.getPath());
+        } catch (final RepositoryException e) {
+            LOGGER.warn("Failed check if form is Survey events form: {}", e.getMessage(), e);
+            return false;
+        }
+    }
+
+    private Node getVisitInformationForm(final Node subject, final Node visitInformationQuestionnaire)
+        throws RepositoryException
+    {
+        final PropertyIterator references = subject.getReferences();
```

Comment:
There's a bug here. If the form that was added was a `Patient information` form, then the subject is a `Patient` subject, and this will just look for the first `Visit information` it finds for the patient and any of its visits. It should be enough to fix it with this minor change:

```suggestion
        final PropertyIterator references = subject.getReferences("subject");
```